### PR TITLE
468-optimize-the-find_local_switch_sequence-function

### PR DESCRIPTION
**Pull request by zlyfunction**
see 
https://github.com/wildmeshing/wildmeshing-toolkit/issues/468

### CODE REVIEW

**Comment by zlyfunction** [comment]
The new implementation is based on https://github.com/wildmeshing/wildmeshing-toolkit/pull/467

File: src/wmtk/multimesh/utils/find_local_switch_sequence.cpp [hunk 1]
```diff
@@ -1,45 +1,132 @@
...
-    const Tuple& target)
+    const Tuple& target,
+    const PrimitiveType mesh_pt)
 {
-    if (source != target) {
-        return std::vector<PrimitiveType>{PrimitiveType::Vertex};
-    } else {
-        return std::vector<PrimitiveType>{};
+    // TODO: test it in the future
+    throw "not tested yet";
```
Comment (mtao) [inline src/wmtk/multimesh/utils/find_local_switch_sequence.cpp:74]: can't just throw a string literal like this. make it a std::runtime_error

File: src/wmtk/multimesh/utils/find_local_switch_sequence.cpp [hunk 2]
```diff
@@ -1,45 +1,132 @@
 #include "find_local_switch_sequence.hpp"
+#include <cassert>
 #include "local_switch_tuple.hpp"
 namespace wmtk::multimesh::utils {
 
 
 namespace {
+std::pair<bool, std::vector<PrimitiveType>> find_local_switch_sequence_on_edge(
```
Comment (mtao) [inline src/wmtk/multimesh/utils/find_local_switch_sequence.cpp:8]: this is on another PR right? you should link / mention this PR in the other one

File: src/wmtk/multimesh/utils/find_local_switch_sequence.cpp [hunk 3]
```diff
@@ -1,45 +1,132 @@
...
+    if (source == target) {
+        return {true, std::vector<PrimitiveType>{}};
+    } else if (local_switch_tuple(mesh_pt, source, PrimitiveType::Vertex) == target) {
+        return {true, std::vector<PrimitiveType>{PrimitiveType::Vertex}};
+    }
+    return {false, std::vector<PrimitiveType>{}};
+}
 
-std::vector<PrimitiveType> find_local_switch_sequence_trimesh(
+std::pair<bool, std::vector<PrimitiveType>> find_local_switch_sequence_on_triangle(
```
Comment (mtao) [inline src/wmtk/multimesh/utils/find_local_switch_sequence.cpp:22]: i prefer std::optional to pairs - that way the user cannot make a mistake

File: src/wmtk/multimesh/utils/find_local_switch_sequence.cpp [hunk 4]
```diff
@@ -1,45 +1,132 @@
...
+    }
+    return {false, std::vector<PrimitiveType>{}};
 }
-std::vector<PrimitiveType> find_local_switch_sequence_edgemesh(
+
+std::pair<bool, std::vector<PrimitiveType>> find_local_switch_sequence_on_tet(
     const Tuple& source,
-    const Tuple& target)
+    const Tuple& target,
+    const PrimitiveType mesh_pt)
```
Comment (mtao) [inline src/wmtk/multimesh/utils/find_local_switch_sequence.cpp:71]: mesh type should be the first argument to satisfy a currying-like semantic

File: src/wmtk/multimesh/utils/find_local_switch_sequence.cpp (hunk 2 above)
Comment (zlyfunction) [reply src/wmtk/multimesh/utils/find_local_switch_sequence.cpp:8]: Ah yes, I mentioned it on the issues linked to that pr.

**Review by JcDai — APPROVED** [review]

**Pull request merged**